### PR TITLE
coord: Apply schema migrations to builtin tables and views

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds the functionality to be able to migrate the schema of
builtin objects. It accomplishes this by dropping all objects that have
changed schema, along with all objects that depend on the original
object. Then it recreates all dropped objects with new GlobalIds.

Previously in the single binary world, schema migrations were
accomplished by restarting the whole system. In the platform world,
each component restarts independently which is why that approach no
longer worked.

Works towards resolving #11435

### Motivation
This PR adds a known-desirable feature.

### Tips for reviewer
* The content of `Catalog::transact` for `Action::DropItem` was moved entirely to `CatalogState::drop_item`. The one change is that `if !id.is_system()` was added to 
```Rust
            if !id.is_system() {
                assert!(
                    self.compute_instances_by_id
                        .get_mut(&compute_instance)
                        .unwrap()
                        .exports
                        .remove(&id),
                    "catalog out of sync"
                );
            }
```
because introspection source indexes are not included in `exports`.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - Manual testing is not possible until command reconciliation is complete. I also don't think we have the testing infrastructure in place to do automated testing for this feature.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - There are no user-facing behavior changes.

### Dependencies

- [X] https://github.com/MaterializeInc/materialize/issues/12096

### CODE REVIEW

File: src/coord/src/catalog.rs
Comment (jkosh44) [inline src/coord/src/catalog.rs:1016]: This whole method feels a little flaky to me, is there a better way to do this?

File: src/coord/src/catalog.rs
Comment (jkosh44) [inline src/coord/src/catalog.rs:1071]: This case feels a little flaky to me, is there a better way to do this?

**Comment by jkosh44** [comment]
@mjibson This PR isn't exactly ready yet, but the changes to storage.rs are done in case you wanted to look.

File: src/coord/src/catalog/storage.rs
Comment (mjibson) [inline src/coord/src/catalog/storage.rs:527]: You need to add this to `commit()` so the updates get persisted.

Comment (jkosh44) [reply src/coord/src/catalog/storage.rs:527]: Thanks, just updated

**Comment by jkosh44** [comment]
One thing that was pointed out by @petrosagg is that if there's some view that's composed of `SELECT * FROM <builtin-table>` and we add a column to that builin-table, then we may have accidentally changed the view definition. An easy fix for this is to have planning convert all `SELECT * ...` into `SELECT <column-list> ...`.

There may be similar scenarios that should be thought through as well.

**Comment by jkosh44** [comment]
Another open question is, if we re-make sinks, then will all the data be re-ingested by the sink?

**Comment by benesch** [comment]
@jkosh44 this should be unblocked now!

File: src/adapter/src/coord.rs
Comment (jkosh44) [inline src/adapter/src/coord.rs:826]: I'm not really a big fan of all these "unvalidated" methods, but I'm not sure what else to do.

**Comment by jkosh44** [comment]
I just realized that this leaks deleted builtins. So I should probably add that.

**Comment by benesch** [comment]
I took a quick skim and this seems super plausible! Unfortunately I don't think I'll have the bandwidth to do a deeper look. Perhaps @mjibson can take a look?

I'm also curious to know if @philip-stoev has any thoughts on how we can test this.

File: src/adapter/src/catalog.rs
Comment (mjibson) [inline src/adapter/src/catalog.rs:2181]: Is it safe to merge with this? Have we not had any of these yet?

**Comment by mjibson** [comment]
Idea for a test: mzcompose.py workflow that starts up some hard coded mz version that is known to need some migrations. The test asserts the ids of a couple of things, shuts it down, spins up a newer mz (this part seems brittle because recreated ids might change over time), then asserts that the ids of those same things have changed. To make this less brittle, it could assert that the recreated ids are just greater than their previous ids, but not care about what the specific ids are.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs:2181]: The other option is to silently ignore them, which is what we currently do. There seems to be ongoing changes to these which is why I left them out for now. I don't think we've had any log migrations since platform, but if we want to add some then we'll have to implement this.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs]: Fixed

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs]: Fixed

**Comment by jkosh44** [comment]
> Idea for a test: mzcompose.py workflow that starts up some hard coded mz version that is known to need some migrations. The test asserts the ids of a couple of things, shuts it down, spins up a newer mz (this part seems brittle because recreated ids might change over time), then asserts that the ids of those same things have changed. To make this less brittle, it could assert that the recreated ids are just greater than their previous ids, but not care about what the specific ids are.
> 
> Maybe it should also include having a user created view and materiaized view over some system objects that will be migrated.

This is a good idea, but I think that since we started working on platform we haven't increased the version number. We also won't be able to upgrade from a pre-platform version to the current version. So I don't think that there exists a valid version to upgrade from to test this.

**Comment by mjibson** [comment]
Yeah, you can specify a specific docker tag with a git commit.

**Comment by benesch** [comment]
We're also (finally) increasing the version numbers again: v0.27.0-alpha.1, v0.27.0-alpha.2, etc.

**Review by philip-stoev — APPROVED** [review]
I wrote a test using the "old Git revision" approach but unfortunately it will not work out -- the revision is apparently so old that the catalog is not preserved when attempting to upgrade from it ... 

I will have to defer adding a test to CI until a future **specific** migration happens between two tagged and supported releases. At that time I will be able to implement the "id-of-the-object-has-increased" idea.

I have also added a checkbox to the Epic around adding a test that **generally** exercises all builtin database objects by creating a materialized view on top of them and making sure the view is readable post-upgrade. Thanks a lot for this idea too.

**Comment by jkosh44** [comment]
Thanks for the help @philip-stoev! I had an idea this morning. I can open a PR that adds one of the pg_catalog tables but omit a column. Then follow it up with a PR that adds in the missing column. We could then use those two commits for testing purposes.

**Comment by jkosh44** [comment]
@philip-stoev Once these two PRs pass CI, I think if we merge them back to back it's OK that we don't have a proper migration. Then we could use those for your tests. Do you think that would work?
- https://github.com/MaterializeInc/materialize/pull/13910
- https://github.com/MaterializeInc/materialize/pull/13915

**Comment by philip-stoev** [comment]
@jkosh44 do not merge them just yet -- once their containers arrive in docker hub I will be able to give it a try without merging them to main. Please give me until tomorrow.

**Comment by philip-stoev** [comment]
I was able to construct a test that seems to do the job. The materialized view created on top of the table that is being migrated survives the migration. Please feel free to push it if you think it is good enough:

```
diff --git a/test/cluster/mzcompose.py b/test/cluster/mzcompose.py
index 35d66a951..b1126b21d 100644
--- a/test/cluster/mzcompose.py
+++ b/test/cluster/mzcompose.py
@@ -8,6 +8,7 @@
 # by the Apache License, Version 2.0.
 
 import time
+from textwrap import dedent
 
 from pg8000.dbapi import ProgrammingError
 
@@ -71,6 +72,7 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
         "test-github-12251",
         "test-remote-storaged",
         "test-drop-default-cluster",
+        "test-schema-migration",
     ]:
         with c.test_case(name):
             c.workflow(name)
@@ -210,3 +212,54 @@ def workflow_test_drop_default_cluster(c: Composition) -> None:
 
     c.sql("DROP CLUSTER default CASCADE")
     c.sql("CREATE CLUSTER default REPLICAS (default (SIZE '1'))")
+
+
+def workflow_test_builtin_migration(c: Composition) -> None:
+    """Exercise the builtin object migration code by upgrading between two versions
+    that will have a migration triggered between them. Create a materialized view
+    over the affected builtin object to confirm that the migration was successful
+    """
+
+    c.down(destroy_volumes=True)
+    with c.override(
+        Testdrive(default_timeout="15s", no_reset=True, consistent_seed=True),
+    ):
+        c.up("testdrive", persistent=True)
+
+        with c.override(
+            Materialized(
+                image="materialize/materialized:devel-ce016efcfd04931e95a2ce6fd90431f68c84804d"
+            )
+        ):
+            c.up("materialized")
+
+            c.testdrive(
+                input=dedent(
+                    """
+            > CREATE MATERIALIZED VIEW v1 AS SELECT COUNT(*) > 0 FROM pg_authid;
+            > CREATE DEFAULT INDEX ON v1;
+        """
+                )
+            )
+
+            c.kill("materialized")
+
+        with c.override(
+            Materialized(
+                image="materialize/materialized:devel-d5328ec05cd391e6b8f16c2e4724f91a15eaa095"
+            )
+        ):
+            c.up("materialized")
+
+            c.testdrive(
+                input=dedent(
+                    """
+       > SELECT * FROM v1;
+       true
+
+       # This column is new after the migration
+       > SELECT DISTINCT rolconnlimit FROM pg_authid;
+       -1
+    """
+                )
+            )
```

**Comment by philip-stoev** [comment]
The test as pasted upgrades to `materialize/materialized:devel-d5328ec05cd391e6b8f16c2e4724f91a15eaa095` but I guess it should upgrade to just `Materialized()` if you are going to add it to the CI. This way the test will work until the "from" image, `ce016efcfd04931e95a2ce6fd90431f68c84804d` becomes too old to upgrade to the latest "main" (which may be a couple of weeks from now, or sooner).

**Comment by jkosh44** [comment]
@mjibson This has a test now which is passing, if you want to re-review whenever you have time.

File: test/cluster/mzcompose.py
Comment (benesch) [inline test/cluster/mzcompose.py:74]: Don't forget to comment these back in!

Comment (jkosh44) [reply test/cluster/mzcompose.py:74]: Oh wow, good catch.

Comment (jkosh44) [reply test/cluster/mzcompose.py]: Fixed

**Review by mjibson — APPROVED** [review]

**Comment by benesch** [comment]
I put this in my queue! Probably won't get to this until this weekend
though, sorry Joe!

**Comment by jkosh44** [comment]
@benesch Did you mean to comment on this?

**Comment by benesch** [comment]
GitHub seems to have found a bunch of undelivered email comments overnight and posted them. So I did mean to comment that but ... like two months ago, lol.

**Pull request merged**
